datastore/search_object.go

```diff
@@ -0,0 +1,106 @@
+package datastore
+
+import (
+	"github.com/PretendoNetwork/nex-go"
+	common_globals "github.com/PretendoNetwork/nex-protocols-common-go/globals"
+	datastore "github.com/PretendoNetwork/nex-protocols-go/datastore"
+	datastore_types "github.com/PretendoNetwork/nex-protocols-go/datastore/types"
+)
+
+func searchObject(err error, client *nex.Client, callID uint32, param *datastore_types.DataStoreSearchParam) uint32 {
+	if commonDataStoreProtocol.getObjectInfosByDataStoreSearchParamHandler == nil {
+		common_globals.Logger.Warning("GetObjectInfosByDataStoreSearchParam not defined")
+		return nex.Errors.Core.NotImplemented
+	}
+
+	if commonDataStoreProtocol.verifyObjectPermissionHandler == nil {
+		common_globals.Logger.Warning("VerifyObjectPermission not defined")
+		return nex.Errors.Core.NotImplemented
+	}
+
+	if err != nil {
+		common_globals.Logger.Error(err.Error())
+		return nex.Errors.DataStore.Unknown
+	}
+
+	// * This is likely game-specific. Also developer note:
+	// * Please keep in mind that no results is allowed. errCode
+	// * should NEVER be DataStore::NotFound!
+	// *
+	// * DataStoreSearchParam contains a ResultRange to limit the
+	// * returned results. TotalCount is the total matching objects
+	// * in the database, whereas objects is the limited results
+	objects, totalCount, errCode := commonDataStoreProtocol.getObjectInfosByDataStoreSearchParamHandler(param)
+	if errCode != 0 {
+		return errCode
+	}
+
+	pSearchResult := datastore_types.NewDataStoreSearchResult()
+
+	pSearchResult.TotalCount = totalCount
+	pSearchResult.Result = make([]*datastore_types.DataStoreMetaInfo, 0, len(objects))
+	pSearchResult.TotalCountType = uint8(param.DataType) // TODO - Is this right?
```

Comment:
No, this represents if all results are returned or not (see https://nintendo-wiki.pretendo.network/docs/nex/protocols/datastore/#datastoresearchresult-structure). It's worth noting that the values on the wiki *could* be flags (0x1: Not all results, 0x2: Disabled), haven't checked though

Reply:
Oh perfect. So I assume that if `len(objects)` matches `totalCount` then `TotalCountType` would be 0, and if not then it would be 1? What is the purpose of option 3 if you know? I'm, unsure of when the client would expect this to be disabled? Or how to even check for that. And do we know if an option 2 exists?

Reply:
> Oh perfect. So I assume that if `len(objects)` matches `totalCount` then `TotalCountType` would be 0, and if not then it would be 1?

Yes, that should be it.

> What is the purpose of option 3 if you know?

The option 3 can be trigegred if the `totalCountEnabled` option in the search params ise set to false. Note that this is only available on revision 3 of the structure or NEX 4.0+

> And do we know if an option 2 exists?

I'm not sure since I haven't seen it

Reply:
I totally missed that field, sweet